src/inspector_io.cc

```diff
@@ -64,407 +58,284 @@ std::string GenerateID() {
   return uuid;
 }
 
-void HandleSyncCloseCb(uv_handle_t* handle) {
-  *static_cast<bool*>(handle->data) = true;
-}
+class RequestQueueData {
+ public:
+  using MessageQueue =
+      std::deque<std::tuple<TransportAction, int,
+                  std::unique_ptr<v8_inspector::StringBuffer>>>;
+
+  explicit RequestQueueData(uv_loop_t* loop):
+      handle_(std::make_shared<RequestQueue>(this)) {
+    int err = uv_async_init(loop, &async_, UvAsyncCb);
+    CHECK_EQ(0, err);
+  }
 
-void CloseAsyncAndLoop(uv_async_t* async) {
-  bool is_closed = false;
-  async->data = &is_closed;
-  uv_close(reinterpret_cast<uv_handle_t*>(async), HandleSyncCloseCb);
-  while (!is_closed)
-    uv_run(async->loop, UV_RUN_ONCE);
-  async->data = nullptr;
-  CheckedUvLoopClose(async->loop);
-}
+  static void CloseAndFree(RequestQueueData* queue) {
+    uv_close(reinterpret_cast<uv_handle_t*>(&queue->async_), FreeOnClose);
+  }
 
-// Delete main_thread_req_ on async handle close
-void ReleasePairOnAsyncClose(uv_handle_t* async) {
-  std::unique_ptr<AsyncAndAgent> pair(node::ContainerOf(&AsyncAndAgent::first,
-      reinterpret_cast<uv_async_t*>(async)));
-  // Unique_ptr goes out of scope here and pointer is deleted.
-}
+  void Post(int session_id,
+            TransportAction action,
+            std::unique_ptr<StringBuffer> message) {
+    Mutex::ScopedLock scoped_lock(state_lock_);
+    bool notify = messages_.empty();
+    messages_.emplace_back(action, session_id, std::move(message));
+    if (notify) {
+      CHECK_EQ(0, uv_async_send(&async_));
+      incoming_message_cond_.Broadcast(scoped_lock);
+    }
+  }
+
+  void Wait() {
+    Mutex::ScopedLock scoped_lock(state_lock_);
+    if (messages_.empty()) {
+      incoming_message_cond_.Wait(scoped_lock);
+    }
+  }
+
+  void SetServer(InspectorSocketServer* server) {
+    server_ = server;
+  }
 
+  std::shared_ptr<RequestQueue> handle() {
+    return handle_;
+  }
+
+ private:
+  ~RequestQueueData();
+
+  static void UvAsyncCb(uv_async_t* async) {
+    RequestQueueData* wrapper =
+        node::ContainerOf(&RequestQueueData::async_, async);
+    wrapper->DoDispatch();
+  }
+
+  static void FreeOnClose(uv_handle_t* handle) {
+    RequestQueueData* wrapper =
+        node::ContainerOf(&RequestQueueData::async_,
+                          reinterpret_cast<uv_async_t*>(handle));
+    delete wrapper;
+  }
```

Comment:
I’d personally make both of these lambdas as well, for readability

Reply:
Done.